File: lib/prompts/extraction.ts
import type { MedicalEntities } from "@/lib/events";

export const EXTRACTION_SYSTEM = `You are a clinical entity extraction agent. Extract structured medical facts from doctor-patient conversation transcripts.

Return ONLY a raw JSON object matching this exact shape (no markdown, no explanation):
{
  "medications": [{ "name": string, "dose"?: string, "frequency"?: string }],
  "conditions": string[],
  "allergies": string[],
  "vitals": { [key: string]: string },
  "symptoms": string[],
  "demographics": { "age"?: number, "sex"?: string }
}

Rules:
- Include every medication, including ones taken long-term (e.g. lisinopril, warfarin)
- Include allergies (e.g. "penicillin allergy")
- Symptoms: include chief complaint and associated symptoms
- Do NOT invent facts not stated in the transcript
- Merge carefully with existing entities — never drop previously extracted data`;

export function buildExtractionPrompt(
  transcript: string,
  existing: MedicalEntities | null
): string {
  return [
    "Existing entities (preserve all, only add new):",
    JSON.stringify(existing ?? {}, null, 2),
    "",
    "New transcript to process:",
    transcript,
    "",
    "Return the fully merged entities JSON.",
  ].join("\n");
}

export function heuristicExtract(
  transcript: string,
  existing: MedicalEntities | null
): MedicalEntities {
  const lower = transcript.toLowerCase();
  const entities: MedicalEntities = existing
    ? JSON.parse(JSON.stringify(existing))
    : { medications: [], conditions: [], allergies: [], vitals: {}, symptoms: [] };

  const addMed = (name: string, dose?: string) => {
    if (!entities.medications.some((m) => m.name.toLowerCase() === name.toLowerCase())) {
      entities.medications.push(dose ? { name, dose } : { name });
    }
  };
  const addCondition = (c: string) => {
    if (!entities.conditions.some((x) => x.toLowerCase() === c.toLowerCase()))
      entities.conditions.push(c);
  };
  const addSymptom = (s: string) => {
    if (!entities.symptoms.some((x) => x.toLowerCase() === s.toLowerCase()))
      entities.symptoms.push(s);
  };
  const addAllergy = (a: string) => {
    if (!entities.allergies.some((x) => x.toLowerCase() === a.toLowerCase()))
      entities.allergies.push(a);
  };

  // Medications
  if (lower.includes("warfarin")) addMed("warfarin");
  if (lower.includes("lisinopril")) addMed("lisinopril");
  if (lower.includes("aspirin")) addMed("aspirin");
  if (lower.includes("metoprolol")) addMed("metoprolol");
  if (lower.includes("atorvastatin") || lower.includes("statin")) addMed("atorvastatin");
  if (lower.includes("heparin")) addMed("heparin");
  if (lower.includes("nitroglycerin") || lower.includes("nitro")) addMed("nitroglycerin");

  // Allergies
  if (lower.includes("penicillin") && (lower.includes("allerg") || lower.includes("reaction")))
    addAllergy("penicillin");
  if (lower.includes("sulfa") && lower.includes("allerg")) addAllergy("sulfa");
  if (lower.includes("nsaid") && lower.includes("allerg")) addAllergy("NSAIDs");

  // Conditions
  if (lower.includes("hypertension") || lower.includes("high blood pressure"))
    addCondition("hypertension");
  if (lower.includes("heart valve")) addCondition("heart valve replacement");
  if (lower.includes("atrial fibrillation") || lower.includes("afib")) addCondition("atrial fibrillation");
  if (lower.includes("diabetes")) addCondition("diabetes mellitus");
  if (lower.includes("heart failure")) addCondition("heart failure");
  if (lower.includes("coronary artery disease") || lower.includes("cad")) addCondition("coronary artery disease");

  // Symptoms
  if (lower.includes("chest pain") || lower.includes("chest pressure")) addSymptom("chest pain");
  if (lower.includes("shortness of breath") || lower.includes("short of breath") || lower.includes("dyspnea"))
    addSymptom("shortness of breath");
  if (lower.includes("left arm")) addSymptom("left arm pain");
  if (lower.includes("jaw pain") || lower.includes("jaw")) addSymptom("jaw pain");
  if (lower.includes("nausea")) addSymptom("nausea");
  if (lower.includes("diaphoresis") || lower.includes("sweating")) addSymptom("diaphoresis");
  if (lower.includes("dizziness") || lower.includes("dizzy")) addSymptom("dizziness");
  if (lower.includes("palpitation")) addSymptom("palpitations");

  // Demographics — age
  const ageMatch = lower.match(/\b(5[5-9]|6[0-9]|7[0-9]|8[0-9])\b/);
  if (ageMatch) {
    entities.demographics = {
      ...entities.demographics,
      age: parseInt(ageMatch[1], 10),
    };
  }
  // Sex
  if (lower.match(/\b(male|man|he|his|gentleman)\b/))
    entities.demographics = { ...entities.demographics, sex: "male" };
  else if (lower.match(/\b(female|woman|she|her|lady)\b/))
    entities.demographics = { ...entities.demographics, sex: "female" };

  return entities;
}

File: backend/prompts/extraction.py
"""
Extraction prompts and heuristic fallback — mirrors lib/prompts/extraction.ts.
"""
from __future__ import annotations

import json
import re
from typing import Optional

from events import MedicalEntities, Medication, Demographics

EXTRACTION_SYSTEM = """You are a clinical entity extraction agent for pre-hospital EMS/ambulance encounters.
Extract structured medical facts from paramedic-patient scene dialogue.

Return ONLY a raw JSON object matching this exact shape (no markdown, no explanation):
{
  "medications": [{ "name": string, "dose"?: string, "frequency"?: string }],
  "conditions": string[],
  "allergies": string[],
  "vitals": { [key: string]: string },
  "symptoms": string[],
  "demographics": { "age"?: number, "sex"?: string }
}

Rules:
- ALLERGIES ARE MANDATORY: Extract all stated drug/substance allergies immediately when mentioned
- Include every medication (including maintenance meds: lisinopril, warfarin, metoprolol, etc.)
- Symptoms: include chief complaint, onset, duration, location, quality, radiation, and associated symptoms
- Use ambulance context: paramedic, patient, bystander, scene — not ER/hospital terminology
- Do NOT invent facts — only extract what was explicitly stated on scene
- Idempotent merge: never drop previously extracted facts; only add new ones
- Vitals: capture if spoken (BP, HR, SpO2, RR, GCS, temperature)"""


def build_extraction_prompt(transcript: str, existing: Optional[MedicalEntities]) -> str:
    from events import to_dict
    existing_dict = to_dict(existing) if existing else {}
    return "\n".join([
        "Existing entities (preserve all — never drop allergies or medications):",
        json.dumps(existing_dict, indent=2),
        "",
        "New scene dialogue to process:",
        transcript,
        "",
        "IMPORTANT: If allergies are mentioned, they MUST appear in the output.",
        "Return the fully merged entities JSON.",
    ])


def heuristic_extract(transcript: str, existing: Optional[MedicalEntities]) -> MedicalEntities:
    from events import to_dict, entities_from_dict
    import copy

    lower = transcript.lower()
    if existing:
        from events import to_dict, entities_from_dict
        entities = entities_from_dict(to_dict(existing))
    else:
        entities = MedicalEntities()

    def add_med(name: str, dose: Optional[str] = None) -> None:
        if not any(m.name.lower() == name.lower() for m in entities.medications):
            entities.medications.append(Medication(name=name, dose=dose))

    def add_condition(c: str) -> None:
        if not any(x.lower() == c.lower() for x in entities.conditions):
            entities.conditions.append(c)

    def add_symptom(s: str) -> None:
        if not any(x.lower() == s.lower() for x in entities.symptoms):
            entities.symptoms.append(s)

    def add_allergy(a: str) -> None:
        if not any(x.lower() == a.lower() for x in entities.allergies):
            entities.allergies.append(a)

    # Medications
    if "warfarin" in lower:
        add_med("warfarin")
    if "lisinopril" in lower:
        add_med("lisinopril")
    if "aspirin" in lower:
        add_med("aspirin")
    if "metoprolol" in lower:
        add_med("metoprolol")
    if "atorvastatin" in lower or "statin" in lower:
        add_med("atorvastatin")
    if "heparin" in lower:
        add_med("heparin")
    if "nitroglycerin" in lower or "nitro" in lower:
        add_med("nitroglycerin")

    # Allergies
    if "penicillin" in lower and ("allerg" in lower or "reaction" in lower):
        add_allergy("penicillin")
    if "sulfa" in lower and "allerg" in lower:
        add_allergy("sulfa")
    if "nsaid" in lower and "allerg" in lower:
        add_allergy("NSAIDs")

    # Conditions
    if "hypertension" in lower or "high blood pressure" in lower:
        add_condition("hypertension")
    if "heart valve" in lower:
        add_condition("heart valve replacement")
    if "atrial fibrillation" in lower or "afib" in lower:
        add_condition("atrial fibrillation")
    if "diabetes" in lower:
        add_condition("diabetes mellitus")
    if "heart failure" in lower:
        add_condition("heart failure")
    if "coronary artery disease" in lower or " cad " in lower:
        add_condition("coronary artery disease")

    # Symptoms
    if "chest pain" in lower or "chest pressure" in lower:
        add_symptom("chest pain")
    if "shortness of breath" in lower or "short of breath" in lower or "dyspnea" in lower:
        add_symptom("shortness of breath")
    if "left arm" in lower:
        add_symptom("left arm pain")
    if "jaw pain" in lower or "jaw" in lower:
        add_symptom("jaw pain")
    if "nausea" in lower:
        add_symptom("nausea")
    if "diaphoresis" in lower or "sweating" in lower:
        add_symptom("diaphoresis")
    if "dizziness" in lower or "dizzy" in lower:
        add_symptom("dizziness")
    if "palpitation" in lower:
        add_symptom("palpitations")

    # Demographics — age
    age_match = re.search(r"\b(5[5-9]|6[0-9]|7[0-9]|8[0-9])\b", lower)
    demo = entities.demographics or Demographics()
    if age_match:
        demo.age = int(age_match.group(1))
    if re.search(r"\b(male|man|he|his|gentleman)\b", lower):
        demo.sex = "male"
    elif re.search(r"\b(female|woman|she|her|lady)\b", lower):
        demo.sex = "female"
    if demo.age or demo.sex:
        entities.demographics = demo

    return entities
